### ISSUE TITLE
Schema datatypes duplicating XSD datatypes considered.

### ISSUE THREAD

**Issue by VladimirAlexiev** [schemaorg/schemaorg#1781]
This issue is related to #1404, #1748 .

#1715 lists https://goo.gl/5TPDw4 (by @ericprud and @labra) which has  @danbri's implicit endorsement. And it shows:
```ttl
     schema:description "A cool dataset"^^schema:Text ;
     schema:url "http://an.url.com"^^schema:URL ;
```

God gave us XSD datatypes to use and cherish, so it gets me really worried if Schema promotes its own unnecessary data types.
- It's fine to say that most props can carry either a structured resource, or `Text` (that's part of Schema's flexibility), but it's not ok to suggest literals should be `^^schema:Text` rather than plain literals (which RDF 1.1 makes the same as `^^xsd:string`).
- As for `URL` they should be URLs and not literals (I could never grok the utility of `xsd:anyURI`)

What's wrong with inventing your own types rather than using XSD types?
- first, it suggests disregard for prior art, which is just wrong
- second, most RDF repos have special handling of numeric and datetime types. Eg GraphDB puts them in a literal index for fast searching, repos know how to do numeric operations, etc. If you declare your own datatypes `owl:equivalentClass` to well-known XSD types, perhaps **D-Entailment** would permit such special handling, but I know of no repo to implement this.

In addition, binding props like modelDate, productionDate, availabilityStarts to datatypes of specific granularity like `schema:Date` and `schema:DateTime` is harmful because it doesn't always reflect the granularity of available data. In #1748 @RichardWallis argues (imho tenaciously) that these also permit other granularities (eg mere year, which corresponds to `xsd:gYear`) and are basically the same thing.

Suggested actions:
1. explain that most Schema datatypes have only organizational purpose but should not be used in actual data
2. document a mapping to XSD as requested by #1404
3. recommend that actual data should be tagged with the most specific XSD datatype, as per [this comment](https://github.com/schemaorg/schemaorg/issues/1748#issuecomment-333153156)
4. make `schema:DateTime` a union of the 5 related XSD datatypes (see the comment) and kill `schema:Date`. Not sure how to make a "union", probably just explain it verbally
5. remove any datatypes from schema's JSONLD context to ensure that data is not mis-interpreted (see  #1748). Map `URL` to `@id`
6. review all examples and make sure they use XSD datatypes

**Comment by RichardWallis** [comment]
@VladimirAlexiev  From the suggested actions in your well argued comments I think there would be benefit gained in adopting:

2. Where there is a direct mapping between a schema datatype and XSD it should be explicitly documented both in supporting RDF files and in user documentation. (#1404)

4. In addition for `schema:DateTime` somehow (I don't know how either) define it as being a union of `xsd:gYear`, `xsd:gYearMonth`, `xsd:gYearMonth`, `xsd:dateTime`,` xsd:dateTimeStamp`.  This would help reflect the practice where many data publishers insert their own interpretation of date format (hopefully something like ISO 8601).  For example I am aware of hundreds of millions of bibliographic descriptions that _mostly_ just provide a 4 digit year for `dateCreated`. 

I believe that removing the `Date` datatype, which has been in the fundamental core of the vocabulary from the beginning, would be very disruptive.  As compromise we could contemplate making `Date` a subtype of `DateTime`.

Your suggestions about `URL` highlights a broader issue where Schema is a victim of its own flexibility.

It is implicit that any property, in addition to its documented expected datatypes, can validly be assigned a value that is `Text` or `URL` or `Role`.  The challenge is how to make this explicit in the vocabulary definition files, especially in the JSON-LD context.  In addressing this problem we need to also agree as to if we need to map `URL` to `@id`.

As to suggestion 6.  
"_Reviewing all examples and make sure..._" is a ambition that appears, or should do, in many requests and proposals for changes and enhancements to the vocabulary.  We have a need to review all current and new examples to check for syntactactical correctness, conformance with the vocabulary at its current release, and reflection of best practices found in documentation published on the Schema.org site.  A challenge is how we satisfy that need  - _volunteers take one step forward!_ 

However, that is not is not a reason for suggesting that we undertake such work.

**Comment by danbri** [comment]
A few quick points -

* Firstly the quick shex example reported in #1715 was thrown together very quickly over lunch yesterday at ISWC, primarily by Iovka and Jose. It is not intended to be the final word on anything. We were trying a few variations to see whether Google validation rules for [dataset descriptions](https://developers.google.com/search/docs/data-types/datasets) might usefully be captured in Shex. One rule was to check that urls were urls. We also btw tried using url-checking regex in @labra) 's system but hit some kind of bug.
* Yes, absolutely we could and should clean up schema.org's datatypes, in particular you are correct that Date and DateTime.
* At launch, schema.org targeted Microdata/HTML5 which was from the browser / web platform world where XML dependencies were generally unwelcome. Since then we have adapted gradually to fit with RDFa and JSON-LD too. 
* There is something special (and awkward) around our use of "URL". We have used it in schemas to hint that a property's values may be links rather than locally described entities, and indeed it is not "Text". Some of this comes from the attempt to embrace both Microdata and RDF-based data models.
* The "Considered harmful" rhetoric doesn't help advance things (but your other comments do!)

**Comment by danbri** [comment]
Here's a version that checks only for some kind of literal, https://goo.gl/EJWdLQ

See also https://goo.gl/pLVyCn from @ericprud

**Comment by VladimirAlexiev** [comment]
hi @RichardWallis!

> need to also agree as to if we need to map `URL` to `@id`

For props that accept a union of types (eg `acceptsReservations`: URL, Boolean or string "Yes"/"No") the context should just be silent.

For props that demand a URL (eg `downloadUrl`), this can only do harm if the data provider misuses the prop, eg:
```ttl
downloadUrl "example.com/downloads" # missing http://
downloadUrl "http://example.com/downloads, last accessed 10 Oct 2017" # random note added
```
Forcing it to `@id` is actually good because downstream software can discover these invalid URLs, which it couldn't do if they were left mere strings.

> volunteers take one step forward!

I am stepping forward, but you guys should organize the work. Eg
- make a sheet where people can signup for specific examples,
- write review guidelines,
- somehow expose the comments or diffs so others can learn from them and complete the work (github PR reviews are a great way)

Hi @danbri!

Those 3 guys are super nice, I reviewed their book extensively, and @ericprud is notoriously attentive to detail: that's why I get worried when they use datatype `^^schema:URL`. If you can vouch that Schema doesn't lead people into using such datatypes then let's close this issue.

>version that checks only for some kind of literal
```
so:url . ;
```
But I think the better way is to check for URL
```
so:url IRI ;
```

No disrespect meant! https://en.wikipedia.org/wiki/Considered_harmful is a sort of "established phrase"; maybe I don't understand the history and I'm not using it correctly? (Just don't say I'm comparing myself to Dijkstra :-))

**Comment by danbri** [comment]
Totally vouch that it was just a quick lunchtime hack! Am about to board a flight, but let's keep this discussion moving and issue open. I think shex/shacl should be the stimulus we need to bring a bit more rigour to schema.org datatyping, and thanks for the nudge :)

**Comment by rvguha** [comment]
"Considered harmful" is about established and helpful as fatty food and ...
i.e., it is old, but not very healthy.

Re XSD data types. There was extensive discussion, before the launch of the
first version of Schema.org, on data types. No, we are not ignoring the ton
of work that has been done on data types over the last 60 years. We decided
that when appropriate, we would use earlier work, giving credit, without
getting into the whole namespace recital requirement.

Yes, lots of schema.org usage is sloppy, like html. We took the decision, I
stand by it, to put the onus on consumers of the data, not the publishers.

guha

**Comment by VladimirAlexiev** [comment]
@rvguha, could you please clarify your position? Do you think publishers should use Schema datatypes or XSD datatypes?

**Comment by danbri** [comment]
@VladimirAlexiev question for you: if schema.org types were annotated with links to xsd where mappings made sense, would Ontotext's database / tooling be able to make more sense of schema.org data?

**Comment by VladimirAlexiev** [comment]
>>owl:equivalentClass to well-known XSD types, perhaps D-Entailment would permit such special handling, but I know of no repo to implement this.
> if schema.org types were annotated with links to xsd where mappings made sense, would Ontotext's database / tooling be able to make more sense

The GraphDB LiteralIndex indexes numbers, dates, strings. The check for dates is `org.eclipse.rdf4j.model.datatypes.XMLDatatypeUtil.isCalendarDatatype(..)`, and that includes the XSD types.

So no, we wouldn't index extra datatypes that are made `owl:equivalentClass` to the well-known XSD types.

**Comment by rvguha** [comment]
Vladimir,

 Our (Schema.org's) goal is to optimize for publishers. We will do whatever
makes things easier for them. There are far fewer programmatic consumers of
the data and they can do the extra work.

 So, should we prefer a few (at most a few dozen) consumers/tool vendors
changing their code or millions of publishers changing their practice?

guha

**Comment by thadguidry** [comment]
@rvguha Let's get millions of publishers to change their practice...and use Schema.org.  Oh wait, we did that already. :)  sorry, just had to. :)  

Although Vladimir's intentions might make us shudder a bit, it is not completely far fetched that publishers like to see well defined standards that provide some value.  This XSD case is certainly not one of them, I agree Guha, but I am also aware that we did get millions of publishers to change their practice once before and in certain cases could probably do it again if it provided really good value.

Agree with @danbri the better way to solve this is getting some XSD mapping in place on our end to help the tooling.

**Comment by VladimirAlexiev** [comment]
@rvguha Do you have actual evidence that Schema datatypes are used widely in published data?
I really hope that's not the case: as I said, I am not aware of any repository to implement D-Entailment, probably because that's not easy.

> We will do whatever makes things easier for them.

Can you explain how using Schema datatypes is easier than using XSD datatypes? Also, what value does that bring? According to @RichardWallis, Schema datatypes `Date` and `DateTime` are vague because they don't indicate the granularity of the actual literal, and the two are in fact the same thing.

@thadguidry and @danbri You seem to express negative opinions about XSD datatypes above, but could you elaborate what is this based upon? What's wrong with XSD datatypes?

@thadguidry You also seem to question my intentions. I assure you they are pure :-) I think we all should use what has been in use for many years and not make new stuff where the existing stuff is perfectly adequate. By the same token, I'm trying to convince the FIBO people that making their own datatypes is folly.

(Don't get me wrong, I'm a big fan of Schema and I really like some of the innovations. Eg I think rdf:domain/range are largely broken since they are monomorphic, and I prefer schema:domain/rangeIncludes)

**Comment by danbri** [comment]
@VladimirAlexiev I am sure your intentions are pure.

I have no objection to XSD. I don't expect for most uses of schema.org, that actual instance data records need explicit datatyping. The information ought to be in the schemas. If there are cases where we have ambiguity or confusion in the schema.org schemas those should be fixed, but I would resist the idea that webmasters/publishers ought to be putting xsd:blablah in millions of pages.

**Comment by gkellogg** [comment]
Gregg Kellogg
gregg@greggkellogg.net

> On Oct 30, 2017, at 7:05 AM, R.V.Guha <notifications@github.com> wrote:
> 
> Vladimir,
> 
> Our (Schema.org's) goal is to optimize for publishers. We will do whatever
> makes things easier for them. There are far fewer programmatic consumers of
> the data and they can do the extra work.
> 
> So, should we prefer a few (at most a few dozen) consumers/tool vendors
> changing their code or millions of publishers changing their practice?
> 
> guha

Note that RDFa and JSON-LD will automatically use XSD datatypes when parsing certain values (@datetype, values of meter and data elements, as well as native JSON values. Unless someone specifically types these using schema datatypes, typed literals with XSD datatypes will be emitted from conforming processors.

Gregg

**Comment by VladimirAlexiev** [comment]
> don't expect for most uses of schema.org, that actual instance data records need explicit datatyping.

Agree. And this issue doesn't say that. 
It just says people should not be misled into using Schema datatypes: which are a nice organizational device, but should recommend the use of "native" datatypes.

> The information ought to be in the schemas 

That's not always possible, given the native flexibility/polymorphism of Schema. As I say above "remove any datatypes from schema's JSONLD context to ensure that data is not mis-interpreted (see #1748)". That issue shows the harm of over-committing certain properties to a specific dateTime granularity.

But when it makes sense, it should be done. Eg `schema:url` and `schema:downloadUrl` should be declared `@id` in the context.

>Unless someone specifically types these using schema datatypes, typed literals with XSD datatypes will be emitted

Thanks for pointing this out, @gkellogg! It shows that previous efforts treat XSD datatypes specially, and IMHO Schema should do the same. I think that XSD types together with IANA lang tags, URLs and Unicode, form the foundation of linked data.

**Comment by danbri** [comment]
Ok, I don't think we have any disagreement here then. Shall we close this issue and follow up in #1748 on specifics?

**Comment by VladimirAlexiev** [comment]
This is a wider issue than #1784 since it's about all Schema types. 
```
Boolean (False, True)
Date
DateTime
Time
Number
  Float
  Integer
Text
  URL
```
See the items 1-6 in the description.

**Comment by VladimirAlexiev** [comment]
Hi Schema folk! Any progress on this issue? Should I start proposing PRs for the datetime datatypes?

**Comment by danbri** [comment]
If there are obvious mappings we can express, just post a list here

**Comment by VladimirAlexiev** [comment]
Mappings schema->xsd datatypes are fairly obvious.
JSONLD context mappings of specific properties are harder because they are applicable only to props with strict range.
See items 1..6 in the description of what I think needs to be done.

schema    | xsd   
----------|-------
Boolean   |boolean
Date      |union gYear, gYearMonth, date, dateTime
DateTime  |union gYear, gYearMonth, date, dateTime
Time      |time
Number    |decimal
  Float   |double
  Integer |integer
Text      |none (IMHO better than string)
  URL     |none, use `@id` in jsonld context

**Comment by ericprud** [comment]
tx for the reminder stale-bot!

@rvguha, where @VladimirAlexiev's mappings are 1:1, how does schema.org benefit from having its own types?

Imo, if folks are consuming them as JSON, then the datatype association in the @context wouldn't matter to them. If they're using SemWeb tools, then having e.g. Turtle and SPARQL support for most of the datatypes is well better than support for none. Union datatypes are effectively duck-typing (which puts the burden on the consumer) but at least the union is closed and some regexp work can turn them into types which can be compared in SPARQL.

**Comment by danbri** [comment]
Can we come up with some examples of Schema.org with its datatypes mapped
to these, such that specific SPARQL engines can answer more
datatype-oriented questions? e.g. with GraphDB what kinds of queries would
currently fail.

Anyone using schema.org data from "in the wild" in practice has to do
various kinds of canonicalization (e.g. http vs https triples), and turning
strings into things. We could document such a transformation at that step
at least, and give sample code.

**Comment by ericprud** [comment]
I think github-actions is trying to remind us to figure this out. If I understand, we're looking for uses of the schema.org analogs of XSD datatypes. Then we examine their use cases to see if it would matter to them if the schema.org @context bound those predicates to `^^schema:X` instead of `^^xsd:X` in a way that motivates the existence of `^^schema:X`.

**Comment by VladimirAlexiev** [comment]
https://github.com/ESIPFed/science-on-schema.org/blob/master/validation/shapegraphs/soso_common_v1.2.2.ttl shows examples of improper expectations about literals:
- sh:class SO:URL
- sh:class SO:Number
- sh:class SO:Text

The proper word is `sh:datatype`.
But as argued above, (I hope that) nobody ever uses eg `"3.14"^^s:Number`

**Comment by jmkeil** [comment]
Hi, looks like there is still no progress about this. Maybe some statistics (as asked for by @VladimirAlexiev) can help, at least for the numeric datatypes. This is the usage of numeric datatypes (from RDF, OWL, Schema.org) in

* the embedded JSON-LD part of the [Web Data Commons (October 2021)](http://webdatacommons.org/structureddata/2021-12/stats/stats.html)
  | Datatype      | Usage absolute | Usage relative to all lexicals |
  | ------------- | -------------- | ------------------------------ |
  | `xsd:integer` | 2593727914     | 0.18                           |
  | `xsd:double`  | 150589610      | 0.01                           |
  | `xsd:float`   | 96972          | 0.00                           |
  | `xsd:decimal` | 27             | 0.00                           |

* the RDFa part of the [Web Data Commons (October 2021)](http://webdatacommons.org/structureddata/2021-12/stats/stats.html)
  | Datatype      | Usage absolute | Usage relative to all lexicals |
  | ------------- | -------------- | ------------------------------ |
  | `xsd:float`   | 261955         | 0.00                           |
  | `xsd:int`     | 2528           | 0.00                           |
  | `xsd:decimal` | 1165           | 0.00                           |

* the embedded JSON-LD part of the [Web Data Commons (December 2020)](http://webdatacommons.org/structureddata/2020-12/stats/stats.html)
  | Datatype      | Usage absolute | Usage relative to all lexicals |
  | ------------- | -------------- | ------------------------------ |
  | `xsd:double`  | 101959406      | 0.01                           |
  | `xsd:float`   | 90895          | 0.00                           |
  | `xsd:long`    | 36             | 0.00                           |
  | `xsd:decimal` | 1              | 0.00                           |

* the RDFa part of the [Web Data Commons (December 2020)](http://webdatacommons.org/structureddata/2020-12/stats/stats.html)
  | Datatype                 | Usage absolute | Usage relative to all lexicals |
  | ------------------------ | -------------- | ------------------------------ |
  | `xsd:float`              | 1025753        | 0.00                           |
  | `xsd:int`                | 19427          | 0.00                           |
  | `xsd:decimal`            | 8288           | 0.00                           |
  | `xsd:nonNegativeInteger` | 586            | 0.00                           |
  | `xsd:double`             | 234            | 0.00                           |
  | `xsd:positiveInteger`    | 6              | 0.00                           |

Up to these data, there is Zero use of `schema:Number`, `schema:Integer` and `schema:Float`. Due to the shorthand syntax for `xsd:integer` and `xsd:double` in JSON-LD (as mentioned by @gkellogg), the use of `xsd:integer` and `xsd:double` is the easiest and people seem to prefer that in contradiction to the schema.org specifications.

As it is the goal to optimize for publishers (as stated by @rvguha), replacing the schema.org numeric datatype with XSD datatypes would be straight forward, preferable before someone minds to start using `schema:Number`, `schema:Integer` or `schema:Float`. This also does not bring further burden with prefixes to publishers: In JSON-LD there are [shorthand syntaxes](https://www.w3.org/TR/json-ld11/#conversion-of-native-data-types), in RDFa there is a [default prefix](https://www.w3.org/TR/rdfa-primer/#default_prefixes) for XSD.

There is a complete different situation with the temporal datatypes, probably due to absence of shorthand syntaxes.  This is the usage of temporal datatypes (from RDF, Schema.org) in

* the embedded JSON-LD part of the [Web Data Commons (October 2021)](http://webdatacommons.org/structureddata/2021-12/stats/stats.html)
  | Datatype          | Usage absolute | Usage relative to all lexicals |
  | ----------------- | -------------- | ------------------------------ |
  | `schema:Date`     | 1384640759     | 0.09                           |
  | `schema:DateTime` | 24827352       | 0.00                           |
  | `xsd:dateTime`    | 40916          | 0.00                           |
  | `xsd:date`        | 7242           | 0.00                           |
  | `xsd:time`        | 2254           | 0.00                           |
  | `xsd:gYear`       | 528            | 0.00                           |
  | `xsd:gYearMonth`  | 95             | 0.00                           |
  | `xsd:duration`    | 14             | 0.00                           |

* the RDFa part of the [Web Data Commons (October 2021)](http://webdatacommons.org/structureddata/2021-12/stats/stats.html)
  | Datatype       | Usage absolute | Usage relative to all lexicals |
  | -------------- | -------------- | ------------------------------ |
  | `xsd:dateTime` | 19355530       | 0.00                           |
  | `xsd:time`     | 32406          | 0.00                           |
  | `xsd:date`     | 21835          | 0.00                           |
  | `xsd:duration` | 2434           | 0.00                           |
  | `xsd:gYear`    | 37             | 0.00                           |

* the embedded JSON-LD part of the [Web Data Commons (December 2020)](http://webdatacommons.org/structureddata/2020-12/stats/stats.html)
  | Datatype          | Usage absolute | Usage relative to all lexicals |
  | ----------------- | -------------- | ------------------------------ |
  | `schema:Date`     | 1313408439     | 0.09                           |
  | `schema:DateTime` | 25002464       | 0.00                           |
  | `xsd:dateTime`    | 12260          | 0.00                           |
  | `xsd:date`        | 3398           | 0.00                           |
  | `xsd:time`        | 1428           | 0.00                           |
  | `xsd:gYear`       | 34             | 0.00                           |
  | `xsd:gYearMonth`  | 20             | 0.00                           |
  | `xsd:duration`    | 2              | 0.00                           |

* the RDFa part of the [Web Data Commons (December 2020)](http://webdatacommons.org/structureddata/2020-12/stats/stats.html)
  | Datatype          | Usage absolute | Usage relative to all lexicals |
  | ----------------- | -------------- | ------------------------------ |
  | `xsd:dateTime`    | 102987223      | 0.02                           |
  | `xsd:date`        | 507454         | 0.00                           |
  | `schema:Date`     | 246995         | 0.00                           |
  | `xsd:time`        | 197426         | 0.00                           |
  | `xsd:gYear`       | 26911          | 0.00                           |
  | `xsd:duration`    | 835            | 0.00                           |
  | `xsd:gYearMonth`  | 58             | 0.00                           |
  | `schema:DateTime` | 7              | 0.00                           |

Here there is a clear preference for `schema:Date` and `schema:DateTime` in embedded JSON-LD and for XSD temporal datatypes in RDFa. The usage of schema.org temporal datatypes in RDFa dropped from low to Zero last year, maybe because of switch to JSON-LD (just speculation)? Interestingly there is Zero use of `schema:Time`.

The numbers base on https://doi.org/10.5281/zenodo.6337661 and https://doi.org/10.5281/zenodo.6205111, analysed with https://gist.github.com/jmkeil/3ef8516183cfd15232d0c84547da369e, which also contain statistics for further datatypes.

**Comment by VladimirAlexiev** [comment]
I tested this query in GraphDB and http://sparql.org/sparql.html (which is Jena ARQ). Would be great if others test it in other  repositories:
```sparql
PREFIX schema: <http://schema.org/>
PREFIX xsd:    <http://www.w3.org/2001/XMLSchema#>
select * {
    bind("2022-01-01"^^schema:Date < "02022-01-02"^^schema:Date as ?x1)
    bind(str("2022-01-01"^^schema:Date) < str("02022-01-02"^^schema:Date) as ?x2)
    bind("2022-01-01"^^xsd:date < "02022-01-02"^^xsd:date as ?x3)
    bind("1"^^schema:Integer < "02"^^schema:Integer as ?x4)
    bind(str("1"^^schema:Integer) < str("02"^^schema:Integer) as ?x5)
    bind("1"^^xsd:integer < "02"^^xsd:integer as ?x6)
}
```
Jena returns the same as GraphDB except `?x3` which is undefined. `arq` returns this, I'll post a bug:
```
WARN  NodeValue       :: Datatype format exception: "02022-01-02"^^xsd:date
```

The conclusions are as expected:
- `schema:Date` and `schema:Integer` are not comparable (return UNDEF) because they are not in SPARQL's operator map
- Taking `str()` of such values makes incorrect comparisons in many cases.
  - `schema:Date` asks  for ISO 8601 dates and https://en.wikipedia.org/wiki/ISO_8601#Years demands at least 4 digits, so string comparison will work for years `0000..9999`
  - But it won't work for years BC (negative), or years after `10000`, or spelled with more digits as my example
  - String comparison of numbers never works (obviously)

@jmkeil @danbri @rvguha @ericprud :
- Sorry to say it, but these several billion `schema:Date` values are screwed up.
- Schema has led data providers into making non-comparable literals.
- The consumer has the burden to cast to `xsd:dateTime`

I also suspect that the granularity of many of these `schema:Date, schema:DateTime` is mis-stated: XSD offers various date-time datatypes of more granularities than Schema.
Blindly casting to `xsd:dateTime` works in GraphDB, eg the following comparisons are correct:

```sparql
    bind("2022"^^xsd:dateTime < "02022-01-02"^^xsd:dateTime as ?x3)
    bind("2022"^^xsd:dateTime = "02022-01-01"^^xsd:dateTime as ?x4)
```

But these comparisons are a non-standard SPARQL extension. They fail on http://sparql.org/sparql.html

**Comment by afs** [comment]
"02022-01-02"^^xsd:dateTime is wrong for two reasons:

1. It starts with a zero but dates more than four digits start 1-9 in XSD. (Provides a canonical representation)
2. It has no time : It is an xsd:date, not an xsd:dateTime. 

Casting would be `xsd:dateTime("-2022-01-02"^^xsd:date)`.

**Comment by VladimirAlexiev** [comment]
@afs you are right on both points, did more testing and checked the spec. Closed https://issues.apache.org/jira/browse/JENA-2318

**Comment by jmkeil** [comment]
What do you (e.g. @danbri, @VladimirAlexiev) think about first harvesting the low hanging fruits, i.e. a potential replacement of the schema numeric datatypes?

Regarding the mapping of `schema:Number` to `xsd:decimal` and `schema:Float` to `xsd:double`, as suggested by @VladimirAlexiev, I see the following problems:
* due to the shorthand syntax, `xsd:double` is heavily used in JSON-LD for properties with the range `schema:Number` (e.g. 49740982 times with `schema:price` in [Web Data Commons (December 2020)](http://webdatacommons.org/structureddata/2020-12/stats/stats.html), see https://doi.org/10.5281/zenodo.6264286)
* `schema:Number` is a superclass of `schema:Float`, but `xsd:decimal` is not a superclass of `xsd:double`, which might result in breaking some logic making use of this relation

Mapping `schema:Number` to `xsd:double` would contradict with `schema:Number` is a superclass of `schema:Integer`.

Even if not nice, mapping `schema:Number` to either `xsd:double` or to both, `xsd:double` and `xsd:decimal`, might better reflect the actual use and trigger fewer problems.

Regarding the temporal datatypes:

I extended [the analysis](https://gist.github.com/jmkeil/3ef8516183cfd15232d0c84547da369e) to show the ratio of valid `xsd:date` and `xsd:dateTime` lexicals per temporal datatype (from RDF, Schema.org) in

* the embedded JSON-LD part of the [Web Data Commons (October 2021)](http://webdatacommons.org/structureddata/2021-12/stats/stats.html)
  |Datatype          |Usage absolute |Usage relative to all lexicals |Valid `xsd:date` notation absolute |Valid `xsd:date` notation ratio |Valid `xsd:dateTime` notation absolute |Valid `xsd:dateTime` notation ratio |Valid `xsd:time` notation absolute |Valid `xsd:time` notation ratio |
  |------------------|---------------|-------------------------------|-----------------------------------|--------------------------------|---------------------------------------|------------------------------------|-----------------------------------|--------------------------------|
  |`schema:Date`     |1384640759     |0.08                           |198639550                          |0.14                            |991794406                              |0.72                                |2585                               |0.00                            |
  |`schema:DateTime` |24827352       |0.00                           |4377660                            |0.18                            |12094933                               |0.49                                |52711                              |0.00                            |
  |`xsd:dateTime`    |40916          |0.00                           |0                                  |0.00                            |40105                                  |0.98                                |0                                  |0.00                            |
  |`xsd:date`        |7242           |0.00                           |7235                               |1.00                            |1                                      |0.00                                |0                                  |0.00                            |
  |`xsd:time`        |2254           |0.00                           |0                                  |0.00                            |0                                      |0.00                                |2254                               |1.00                            |
  |`xsd:gYear`       |528            |0.00                           |0                                  |0.00                            |0                                      |0.00                                |0                                  |0.00                            |
  |`xsd:gYearMonth`  |95             |0.00                           |0                                  |0.00                            |0                                      |0.00                                |0                                  |0.00                            |
  |`xsd:duration`    |14             |0.00                           |0                                  |0.00                            |0                                      |0.00                                |0                                  |0.00                            |

* the RDFa part of the [Web Data Commons (October 2021)](http://webdatacommons.org/structureddata/2021-12/stats/stats.html)
  |Datatype       |Usage absolute |Usage relative to all lexicals |Valid `xsd:date` notation absolute |Valid `xsd:date` notation ratio |Valid `xsd:dateTime` notation absolute |Valid `xsd:dateTime` notation ratio |Valid `xsd:time` notation absolute |Valid `xsd:time` notation ratio |
  |---------------|---------------|-------------------------------|-----------------------------------|--------------------------------|---------------------------------------|------------------------------------|-----------------------------------|--------------------------------|
  |`xsd:dateTime` |19355530       |0.05                           |27505                              |0.00                            |19243315                               |0.99                                |0                                  |0.00                            |
  |`xsd:time`     |32406          |0.00                           |0                                  |0.00                            |0                                      |0.00                                |21334                              |0.66                            |
  |`xsd:date`     |21835          |0.00                           |16811                              |0.77                            |1482                                   |0.07                                |0                                  |0.00                            |
  |`xsd:duration` |2434           |0.00                           |0                                  |0.00                            |0                                      |0.00                                |0                                  |0.00                            |
  |`xsd:gYear`    |37             |0.00                           |0                                  |0.00                            |0                                      |0.00                                |0                                  |0.00                            |

* the embedded JSON-LD part of the [Web Data Commons (December 2020)](http://webdatacommons.org/structureddata/2020-12/stats/stats.html)
  |Datatype          |Usage absolute |Usage relative to all lexicals |Valid `xsd:date` notation absolute |Valid `xsd:date` notation ratio |Valid `xsd:dateTime` notation absolute |Valid `xsd:dateTime` notation ratio |Valid `xsd:time` notation absolute |Valid `xsd:time` notation ratio |
  |------------------|---------------|-------------------------------|-----------------------------------|--------------------------------|---------------------------------------|------------------------------------|-----------------------------------|--------------------------------|
  |`schema:Date`     |1313408439     |0.09                           |176077265                          |0.13                            |967033365                              |0.74                                |1143                               |0.00                            |
  |`schema:DateTime` |25002464       |0.00                           |6069504                            |0.24                            |10580352                               |0.42                                |94628                              |0.00                            |
  |`xsd:dateTime`    |12260          |0.00                           |0                                  |0.00                            |11668                                  |0.95                                |0                                  |0.00                            |
  |`xsd:date`        |3398           |0.00                           |3274                               |0.96                            |0                                      |0.00                                |0                                  |0.00                            |
  |`xsd:time`        |1428           |0.00                           |0                                  |0.00                            |0                                      |0.00                                |1428                               |1.00                            |
  |`xsd:gYear`       |34             |0.00                           |0                                  |0.00                            |0                                      |0.00                                |0                                  |0.00                            |
  |`xsd:gYearMonth`  |20             |0.00                           |0                                  |0.00                            |0                                      |0.00                                |0                                  |0.00                            |
  |`xsd:duration`    |2              |0.00                           |0                                  |0.00                            |0                                      |0.00                                |0                                  |0.00                            |

* the RDFa part of the [Web Data Commons (December 2020)](http://webdatacommons.org/structureddata/2020-12/stats/stats.html)
  |Datatype          |Usage absolute |Usage relative to all lexicals |Valid `xsd:date` notation absolute |Valid `xsd:date` notation ratio |Valid `xsd:dateTime` notation absolute |Valid `xsd:dateTime` notation ratio |Valid `xsd:time` notation absolute |Valid `xsd:time` notation ratio |
  |------------------|---------------|-------------------------------|-----------------------------------|--------------------------------|---------------------------------------|------------------------------------|-----------------------------------|--------------------------------|
  |`xsd:dateTime`    |102987223      |0.02                           |74220                              |0.00                            |102607586                              |1.00                                |0                                  |0.00                            |
  |`xsd:date`        |507454         |0.00                           |367919                             |0.73                            |18905                                  |0.04                                |0                                  |0.00                            |
  |`schema:Date`     |246995         |0.00                           |368                                |0.00                            |246219                                 |1.00                                |0                                  |0.00                            |
  |`xsd:time`        |197426         |0.00                           |0                                  |0.00                            |0                                      |0.00                                |141612                             |0.72                            |
  |`xsd:gYear`       |26911          |0.00                           |0                                  |0.00                            |2396                                   |0.09                                |0                                  |0.00                            |
  |`xsd:duration`    |835            |0.00                           |0                                  |0.00                            |0                                      |0.00                                |0                                  |0.00                            |
  |`xsd:gYearMonth`  |58             |0.00                           |0                                  |0.00                            |28                                     |0.48                                |0                                  |0.00                            |
  |`schema:DateTime` |7              |0.00                           |0                                  |0.00                            |7                                      |1.00                                |0                                  |0.00                            |

This shows that casting `schema:Date` to `xsd:dateTime` is considerable more likely to succeed then casting to `xsd:date`. It is even more likely to succeed then casting `schema:DateTime` to `xsd:dateTime`. I conclude that `schema:DateTime` and `schema:Date` are in practice used synonymously and data consumers must be prepared for every kind of temporal lexical. (This is also related to #1748.)

**Comment by ericprud** [comment]
tldr; me like strongly-typed literals

I work with a clinical data standard called FHIR. The JSON and XML representations have union types that make it easy to author compliant data 'cause they permit e.g. gYear, gYearMonth, date or dateTime. You can throw whatever you've got in there and it will probably be compliant.

It totally sucks to use that data. Every query has to have an enormous amount of lexical analysis built in. (In practice, they don't, so there's lots of opportunity for things to go wrong.) FHIR/RDF requires literals to have the actual datatype, which gives the querier/consumer the necessary control to write sound queries/code.

Emphatic vote for putting the typing burden on the producer.

**Comment by VladimirAlexiev** [comment]
@jmkeil I've reread your great analysis above (as part of writing recommendations for a client to reuse schema.org widely, but not to use Schema datatypes).
Do you want to write a joint paper about this?

**Comment by VladimirAlexiev** [comment]
Slightly related to #1511

**Comment by MidnightLightning** [comment]
Reviving this conversation as I came across this apparent gap between Schema.org datatypes and XSD datatypes when I realized the JSON-LD export that is advertised on https://schema.org/docs/developers.html includes "xsd" as a prefix in the @context block, but then doesn't use that prefix anywhere in the document. If it is the decision of the Schema.org maintainers to not directly link their lowest-level datatypes to XSD datatypes, it would be good to not include that prefix as an alias in the export. As it currently is, I thought it was an export mistake, where there was an intended link and those annotations just didn't make it into the export file.
